src/sqs_consumer/sequential.clj

```diff
@@ -1,27 +1,63 @@
 (ns sqs-consumer.sequential
   (:require [amazonica.aws.sqs :as sqs]
-            [sqs-consumer.core :as core]
-            [sqs-consumer.utils :as utils]))
+            [sqs-consumer.core :as core]))
 
-(defn delete-message [{:keys [queue-url aws-config]} receipt-handle]
+(set! *warn-on-reflection* true)
+
+(defn- delete-message [{:keys [queue-url aws-config]} receipt-handle]
   (sqs/delete-message aws-config queue-url receipt-handle))
 
+(defn- change-message-visibility [{:keys [queue-url aws-config]} receipt-handle visibility-timeout]
+  (sqs/change-message-visibility aws-config queue-url receipt-handle visibility-timeout))
+
+(defn extract-message
+  "Extracts the message from the SQS receive messages response.
+   
+   Each message is an SQS Message https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/index.html?com/amazonaws/services/sqs/model/Message.html."
+  [config message]
+  {;; the original message, with unparsed body.
+   :message message
+   ;; attributes on the message, note these are different to message-attributes, e.g. ApproximateReceiveCount
+   :attributes (:attributes message)
+   ;; the message body itself. If this was the message from SNS it is wrapped in an SNS envelope unless RawMessageDelivery is set to true
+   :message-body (:body message)
+   ;; map of any message attributes set via the SQS send-message API
+   :message-attributes (core/parse-message-attributes (:message-attributes message))
+   ;; the sqs delete message API call pre-bound to the given message/aws config
+   :delete-message #(delete-message config (:receipt-handle message))
+    ;; the sqs change message visibility API call pre-bound to the given message/AWS config
+   :change-message-visibility #(change-message-visibility config (:receipt-handle message) %)})
+
 (defn sequential-process [process-fn]
   (fn [{:keys [config messages]}]
-    (run! (fn [message] (process-fn {:message-body (:body message)
-                                     :delete-message #(delete-message config (:receipt-handle message))}))
+    (run! (fn [message] (process-fn (extract-message config message)))
           messages)))
 
-(defn with-auto-delete [process-fn]
+(defn with-auto-delete
+  "Auto deletes the message from SQS whenever process-fn returns a truthy value.
+   If a non-truthy value is returned or an exception is thrown the message will not be deleted."
+  [process-fn]
   (fn [{:keys [delete-message] :as message}]
     (when (process-fn (dissoc message :delete-message))
       (delete-message))))
 
-(defn with-decoder [process-fn decoder]
-  (fn [{:keys [message]}]
-    (process-fn (decoder message))))
+(defn with-decoder
+  "Merges the message with the result of the given message decoder. The decoder is called with the :message-body from the extract-message function."
+  [process-fn decoder]
+  (fn [message]
+    (-> message
+        (merge (decoder (:message-body message)))
```

Comment:
Thinking on this, the decoder should probably just return the full message map back with modification rather than doing a merge, to prevent having to consider weird deep-merging semantics

Reply:
I'm thinking we should probably just get rid of `with-decoder` and make the utils functions operate on the full map, then they can mutate things however they want.

As long as the payload for each call in the process-fn chain is consistent and documented I think this is saner.

It would also potentially allow us to instead accept a list of middleware, as opposed to a `process-fn`. That has the advantages of being able to do things between the given middleware in the stack, e.g print a debugging log of what was changed by the middleware.

Reply:
So you're suggesting something more like how [Reitit handles middleware](https://github.com/metosin/reitit#ring-example) - where it's an explicit array of functions, rather than a chain of higher order functions?

Reply:
Potentially. It allows the library to help out more in development. Reitit is a good example as I was particularly thinking of this feature there: https://cljdoc.org/d/metosin/reitit/0.5.5/doc/ring/default-middleware#inspecting-middleware-chain

Reply:
Yeah. That makes loads of sense. I've had trouble with that in the past with local dev where there is a lot to keep in your head when making changes to middleware.

Schema checking is where my mind went immediately as well. It would be a logical extension that could make it simpler again to reason about. The ability to say my middleware expects values of this shape and produces values of another shape would be cool.